src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs

```diff
@@ -77,10 +83,22 @@ internal partial class DiagnosticIncrementalAnalyzer
 
                 if (analyzer.IsCompilerAnalyzer())
                 {
-                    // Compiler analyzer is treated specially.
-                    // It is executed for all documents (open and closed) for 'BackgroundAnalysisScope.FullSolution'
-                    // and executed for just open documents for other analysis scopes.
-                    return analysisScope == BackgroundAnalysisScope.FullSolution || isOpenDocument;
+                    return compilerDiagnosticsScope switch
+                    {
+                        // Compiler diagnostics are disabled for all documents.
+                        CompilerDiagnosticsScope.None => false,
+
+                        // Compiler diagnostics are enabled for visible documents and open documents which had errors/warnings in prior snapshot.
+                        CompilerDiagnosticsScope.VisibleFiles => isVisibleDocument || (isOpenDocument && !previousData.Items.IsEmpty),
```

Comment:
internally, i think we should call this flag `VIsibleFilesAndFilesWithPreviouslyReportedDiagnostics`